src/build_system.ml

```diff
@@ -394,81 +400,84 @@ let set_rule_generators ~init ~gen_rules =
   Fdecl.set t.init_rules init_rules;
   Fdecl.set t.gen_rules gen_rules
 
-let get_dir_status t ~dir =
-  match Path.Table.find t.dirs dir with
-  | Some x -> Some x
+let get_dir_triage t ~dir =
+  match Path.as_in_source_tree dir with
+  | Some dir ->
+    Dir_triage.Known (Non_build (
+      Path.Source.Set.to_list (File_tree.files_of t.file_tree dir)
+      |> List.map ~f:Path.source
+      |> Path.Set.of_list))
   | None ->
-    let res =
-      match Path.as_in_source_tree dir with
-      | Some dir ->
-        Some (Dir_status.Loaded {
+    if Path.equal dir Path.build_dir then
+      (* Not allowed to look here *)
+      Dir_triage.Known
+        (Build
+           { rules_produced = Rules.empty;
+             targets_here = Path.Map.empty;
+             targets_of_alias_dir = Path.Map.empty;
+           })
+    else if not (Path.is_managed dir) then
+      Dir_triage.Known
+        (Non_build (
+           match Path.readdir_unsorted dir with
+           | Error Unix.ENOENT -> Path.Set.empty
+           | Error m ->
+             Errors.warn Loc.none
+               "Unable to read %s@.Reason: %s@."
+               (Path.to_string_maybe_quoted dir)
+               (Unix.error_message m);
+             Path.Set.empty
+           | Ok files ->
+             Path.Set.of_list (List.map files ~f:(Path.relative dir))))
+    else begin
+      let (ctx, sub_dir) = Path.extract_build_context_exn dir in
+      if ctx = ".aliases" then
```

Comment:
Would it be possible to use `Utils.analyse_target` here?

Reply:
I think this requires a bit more clean up than I want to put in this PR. Let's delay that until after this is merged?